### PR TITLE
#135 Added schema location to metric.xml

### PR DESCRIPTION
**Pull request by llorllale**
as per #135:

Added `xsi:noNamespaceSchemaLocation` to the generated `metric.xml` and left puzzles for the rest of the schemas in `src/resources/org/jpeek/xsd`

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#196](https://github.com/yegor256/jpeek/issues/196) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#196](https://github.com/yegor256/jpeek/issues/196) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/491415268). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by llorllale** [comment]
@amihaiemil please review

**Comment by llorllale** [comment]
@amihaiemil just fixed some conflicts, please review

File: src/main/java/org/jpeek/App.java [hunk 1]
```diff
@@ -131,7 +131,8 @@ public App(final Path source, final Path target) {
                 new MapEntry<>("LCOM5", true),
                 new MapEntry<>("SCOM", true),
                 new MapEntry<>("NHD", true),
-                new MapEntry<>("MMAC", true)
+                new MapEntry<>("MMAC", true),
+                new MapEntry<>("OCC", true)
```
Comment (amihaiemil) [inline src/main/java/org/jpeek/App.java]: @llorllale Why did you add OCC here? Does it have to do with metric.xsl?

File: src/main/java/org/jpeek/Report.java [hunk 2]
```diff
@@ -130,15 +139,36 @@
...
      */
+    @SuppressWarnings("unchecked")
     Report(final XML xml, final String name,
         final Map<String, Object> args,
         final double mean, final double sigma) {
         this.skeleton = xml;
         this.metric = name;
-        this.params = args;
+        this.params = new MapOf<>(
+            // @checkstyle LineLength (1 line)
```
Comment (amihaiemil) [inline src/main/java/org/jpeek/Report.java]: @llorllale Can you indent this code so you don't have to escape checkstyle?

File: src/main/java/org/jpeek/Report.java [hunk 3]
```diff
@@ -130,15 +139,36 @@
...
+            args, new MapEntry<>("schemaLocation", Report.SCHEMA_FILE)
+        );
         this.post = new XSLChain(
             new CollectionOf<>(
+                new XSLDocument(
+                    new UncheckedText(
+                        new TextOf(
+                            new InputOf(
+                                // @checkstyle LineLength (1 line)
```
Comment (amihaiemil) [inline src/main/java/org/jpeek/Report.java]: @llorllale Same here

File: src/main/resources/org/jpeek/metrics/OCC.xsl [hunk 4]
```diff
@@ -0,0 +1,107 @@
+<?xml version="1.0"?>
```
Comment (amihaiemil) [inline src/main/resources/org/jpeek/metrics/OCC.xsl]: @llorllale Again, I don't understand why this PR adds both OCC metric and implements the metric.xsl thing?

File: src/main/resources/org/jpeek/xsd/index.xsd [hunk 5]
```diff
@@ -22,6 +22,14 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 -->
+<!--
+@todo #134:30min Index.xsd: add `xsd:documentation` tags as per issue #134. We need to
```
Comment (amihaiemil) [inline src/main/resources/org/jpeek/xsd/index.xsd]: @llorllale Why are you adding these todos here? They have nothing to do with the ticket or the ``metric.xsl`` functionality, right?

**Comment by llorllale** [comment]
@amihaiemil fixed, please review.

You were right. Lucky for me however, those other changes are now in `master` so it's all good. :)

File: src/main/java/org/jpeek/Report.java (hunk 3 above)
Comment (llorllale) [reply src/main/java/org/jpeek/Report.java]: @amihaiemil fixed

File: src/main/java/org/jpeek/Report.java (hunk 2 above)
Comment (llorllale) [reply src/main/java/org/jpeek/Report.java]: @amihaiemil fixed

File: src/main/resources/org/jpeek/metrics/OCC.xsl (hunk 4 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/metrics/OCC.xsl]: @amihaiemil I fixed this

File: src/main/resources/org/jpeek/xsd/index.xsd (hunk 5 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/xsd/index.xsd]: @amihaiemil fixed, they had nothing to do with the ticket

**Comment by llorllale** [comment]
@amihaiemil fixed the indentation issues

File: src/main/java/org/jpeek/App.java (hunk 1 above)
Comment (llorllale) [reply src/main/java/org/jpeek/App.java]: @amihaiemil fixed; this was an error on my part

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by amihaiemil** [comment]
@yegor256 ping

File: src/main/java/org/jpeek/Report.java [hunk 6]
```diff
@@ -130,15 +139,37 @@
...
+            new MapEntry<>("schemaLocation", Report.SCHEMA_FILE)
+        );
         this.post = new XSLChain(
             new CollectionOf<>(
+                new XSLDocument(
+                    new UncheckedText(
+                        new TextOf(
+                            new InputOf(
+                                Report.class.getResourceAsStream(
+                                    "xsl/metric-post-schemaloc.xsl"
```
Comment (yegor256) [inline src/main/java/org/jpeek/Report.java]: @llorllale looks weird. Can't we just add this single attribute to the `skeleton.xml` when we are building it? Why making this simple operation so complex?

File: src/main/resources/org/jpeek/xsd/index.xsd [hunk 7]
```diff
@@ -26,6 +26,9 @@ SOFTWARE.
 @todo #134:30min Index.xsd: add `xsd:documentation` tags as per issue #134. We need to
  document the schemas in src/resources/org/jpeek/xsd so that the semantics of
  the generated XML documents is transparent for maintainers.
+@todo #135:30min Index.xsd: add a reference to this schema from the generated
+ skeleton.xml. Schemas from src/resources/org/jpeek/xsd must be referenced in
```
Comment (yegor256) [inline src/main/resources/org/jpeek/xsd/index.xsd]: @llorllale what `skeleton.xml` has to do with this file? Typo?

File: src/main/resources/org/jpeek/xsd/matrix.xsd [hunk 8]
```diff
@@ -26,6 +26,9 @@ SOFTWARE.
 @todo #134:30min Matrix.xsd: add `xsd:documentation` tags as per issue #134. We need to
  document the schemas in src/resources/org/jpeek/xsd so that the semantics of
  the generated XML documents is transparent for maintainers.
+@todo #135:30min Matrix.xsd: add a reference to this schema from the generated
+ skeleton.xml. Schemas from src/resources/org/jpeek/xsd must be referenced in
```
Comment (yegor256) [inline src/main/resources/org/jpeek/xsd/matrix.xsd]: @llorllale what `skeleton.xml` has to do with this file? Typo?

File: src/main/resources/org/jpeek/xsd/skeleton.xsd [hunk 9]
```diff
@@ -22,6 +22,11 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 -->
+<!--
+@todo #135:30min Skeleton.xsd: add a reference to this schema from the generated
+ skeleton.xml. Schemas from src/resources/org/jpeek/xsd must be referenced in
```
Comment (yegor256) [inline src/main/resources/org/jpeek/xsd/skeleton.xsd]: @llorllale what `skeleton.xml` has to do with this file? Typo?

**Comment by yegor256** [comment]
@llorllale it seems you over-designed it, see above

File: src/main/resources/org/jpeek/xsd/index.xsd (hunk 7 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/xsd/index.xsd]: @yegor256 yes, typo

File: src/main/resources/org/jpeek/xsd/matrix.xsd (hunk 8 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/xsd/matrix.xsd]: @yegor256 yes, typo

File: src/main/resources/org/jpeek/xsd/skeleton.xsd (hunk 9 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/xsd/skeleton.xsd]: @yegor256 yes, typo

File: src/main/java/org/jpeek/Report.java (hunk 6 above)
Comment (llorllale) [reply src/main/java/org/jpeek/Report.java]: @yegor256 My line of reasoning is: `Report` creates `metric.xml` so it is responsible for properly building it.

`Skeleton` creates the `skeleton.xml` so I believe it is responsible for properly building that xml.

I see that `App` is currently creating both `index.xml` and `matrix.xml` and I think those two things should be extracted into their own classes who in turn will be responsible for properly building those xmls.

WDYT?

Comment (llorllale) [reply src/main/java/org/jpeek/Report.java]: @yegor256 let's also remember that the value for `xsi:noNamespaceSchemaLocation` will be different among the files.

**Comment by amihaiemil** [comment]
@llorllale Are you done with the changes asked by Yegor? If so, ping him so we can try to merge

**Comment by llorllale** [comment]
Ping @yegor256

**Comment by llorllale** [comment]
@yegor256 any thoughts ?

File: src/main/java/org/jpeek/Report.java (hunk 6 above)
Comment (yegor256) [reply src/main/java/org/jpeek/Report.java]: @llorllale why can't you set it via Xembly? Why using XSL?

**Comment by yegor256** [comment]
@llorllale see above

File: src/main/java/org/jpeek/Report.java (hunk 6 above)
Comment (llorllale) [reply src/main/java/org/jpeek/Report.java]: @yegor256 

> why can't you set it via Xembly?

`Xembler` works with nodes. Can't set the required attribute on the node returned by the `XML` resulting from transforming the skeleton into `metric` because it's a document node (throws error "Cannot cast ...DocumentImpl to ...Element"). I think I would have to bend over backwards in the code to make it work.

> Why using XSL?

We already have a post-processing stage in `Report` so I took advantage of that. Also, this way seems cleaner to me than the alternative I described above. The noise can be reduced a little if `XSLDocument` had a ctor that accepted `InputStream` and `params` together.

**Comment by llorllale** [comment]
@yegor256 please see my reply above. Also, please let me know your thoughts regarding the puzzle I'm leaving here as it also involves a design decision.

**Comment by amihaiemil** [comment]
@yegor256 ping

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@llorllale it still seems weird to me that we are using XSL just in order to add one attribute to the XML. We have to use Xembly, it will take just one instruction and the problem is solved. Current solution is too complex and verbose.

**Comment by llorllale** [comment]
@yegor256 where would this one instruction go? Sorry, for asking, I don't see it.

**Comment by amihaiemil** [comment]
@llorllale @yegor256 If you can't agree on the solution, then maybe you should open another PR and solve it there, since this is in my hands for 8 days now :D

**Comment by llorllale** [comment]
@amihaiemil agreed, should I just close this?

**Comment by amihaiemil** [comment]
@llorllale yes. I mean, I can't think of a better solution for this kind of situation. I, as the CR, did my job, but clock is ticking. I was thinking that now the project will have to pay for a second CR -- but then again, if I would continue to review this one, I would spend more than 15min so it wouldn't be correct.

So yeah, I think this is the best solution, just close this and open another one, discuss further changes there :)

**Comment by llorllale** [comment]
@yegor256 the solution I've provided works and moves the project forward. If an improvement needs to be made then I understand we can leave a puzzle for it. Agree?

This job has gone beyond budget for both DEV and REV.

**Comment by yegor256** [comment]
@llorllale sure, you can always "puzzle" the problem to cut corners and close the ticket.

**Comment by llorllale** [comment]
@amihaiemil @yegor256 I've changed the puzzle; now leaving the refactoring work as the puzzle

**Comment by amihaiemil** [comment]
@yegor256 can we merge this now, it's the 10th day, 0crat will take it away tonight, I think :(

**Comment by 0crat** [comment]
The user @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#196](https://github.com/yegor256/jpeek/issues/196), please stop working. Reason for job resignation: It is older than 10 days, see [§8](http://www.zerocracy.com/policy.html#8)

**Comment by 0crat** [comment]
Resigned on delay, see [§8](http://www.zerocracy.com/policy.html#8): -15 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+1145](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Mistake of @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)[/z](https://www.0crat.com/u/amihaiemil) (your student): Resigned on delay, see [§8](http://www.zerocracy.com/policy.html#8): -15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+11029](https://www.0crat.com/u/yegor256)

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
The job [#196](https://github.com/yegor256/jpeek/issues/196) is now out of scope
